### PR TITLE
Don't save Jenkins configuration on reload

### PR DESCRIPTION
**Pull request by Vlatombe**
`Jenkins#setSecurityRealm` saves Jenkins configuration. When using the "Reload configuration from disk", this method is called so that Servlet filters are set appropriately for the potentially new security realm.
This extracts the corresponding logic in order to avoid calling the wholesale setter, and avoid the unnecessary `save()` call.

### Proposed changelog entries

* Avoid unnecessary configuration save when reloading configuration from disk.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [X] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [X] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7305"></a>

### CODE REVIEW

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (Vlatombe) [inline core/src/main/java/jenkins/model/Jenkins.java:3488]: Replaced by the call in `loadConfig()` which is closer to where the `securityRealm` field is set.

**Comment by daniel-beck** [comment]
IIUC, as a side effect, this will no longer save Jenkins during startup, correct?

**Comment by Vlatombe** [comment]
> this will no longer save Jenkins during startup

Indeed

File: test/src/test/java/jenkins/model/JenkinsTest.java
Comment (jglick) [inline test/src/test/java/jenkins/model/JenkinsTest.java:782]: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/attribute/FileTime.html#equals(java.lang.Object) is this reliable across platforms?

Might be more reliable to use a `SaveableListener` as a `@TestExtension`.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jglick) [reply core/src/main/java/jenkins/model/Jenkins.java:3488]: FTR https://github.com/jenkinsci/jenkins/blob/47c26ade47c51fd705f93a8e97e75a69771cf072/core/src/main/java/jenkins/model/Jenkins.java#L3395-L3410

**Review by jglick — APPROVED** [review]
Looks right.

**Review by basil — APPROVED** [review]
Thanks for the PR!

---

This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Review by NotMyFault — APPROVED** [review]

**Pull request merged**
